pyteal/ast/frame.py

```diff
@@ -11,8 +14,101 @@
     from pyteal.compiler import CompileOptions
 
 
+@dataclass
+class LocalTypeSegment:
+    local_type: TealType
+    cnt: int
+    auto_instance: Expr = field(init=False)
+
+    def __post_init__(self):
+        if self.cnt <= 0:
+            raise TealInternalError(
+                "LocalTypeSegment initialization error: segment length must be strictly greatly than 0."
+            )
+        match self.local_type:
+            case TealType.uint64 | TealType.anytype:
+                self.auto_instance = Int(0)
+            case TealType.bytes:
+                self.auto_instance = Bytes("")
+            case TealType.none:
+                raise TealInternalError(
+                    "Local variable in subroutine initialization must be typed."
+                )
+
+    def __teal__(self, options: "CompileOptions") -> tuple[TealBlock, TealSimpleBlock]:
+        if self.cnt == 1:
+            inst_srt, inst_end = self.auto_instance.__teal__(options)
+            return inst_srt, inst_end
+        else:
+            dupn_srt, dupn_end = DupN(self.auto_instance, self.cnt).__teal__(options)
+            return dupn_srt, dupn_end
+
+
+LocalTypeSegment.__module__ = "pyteal"
+
+
+@dataclass
+class ProtoStackLayout:
+    arg_stack_types: list[TealType]
+    local_stack_types: list[TealType]
+    has_output: bool
+    output_index: Optional[int] = field(init=False)
+    succinct_repr: list[LocalTypeSegment] = field(init=False)
+
+    def __post_init__(self):
+        if self.has_output and len(self.local_stack_types) == 0:
+            raise TealInternalError(
+                "ProtoStackLayout initialization error: cannot output without local variable allocs."
+            )
+
+        self.output_index = 0 if self.has_output else None
+        if not all(map(lambda t: types_match(t, TealType.none), self.arg_stack_types)):
+            raise TealInternalError("Variables in frame memory layout must be typed.")
+
+        # Type check of local variables are performed over LocalTypeSegments
+        self.succinct_repr = [
+            LocalTypeSegment(t_type, len(list(dup_seg)))
+            for t_type, dup_seg in groupby(self.local_stack_types)
+        ]
+
+        # anytype at i + 1 merge to i
+        for i in reversed(range(len(self.succinct_repr) - 1)):
```

Comment:
```suggestion
        for i in range(len(self.succinct_repr) - 2, -1, -1)):
```

not sure this is actually better

Reply:
resolving for I removed the whole algorithm...